joomla/joomla-cms
# [4.0][RFC] Child templates a-la wp, but better

# First and foremost what is proposed here is totally B/C and opt-in

### The problem
Let's do a simple case study:
Imagine that you're a dev/designer/admin/etc and you decided to use a 3rd PD template. To make this more realistic let's say that you want to use @C-Lodder https://github.com/C-Lodder/joomla4-backend-template 
Few days later you decide that some changes in one of the overrides will fit better your use case. You ego ahead and do the changes. Couple days later Joomla says that the template has an update and you go ahead and apply the update. 

**Ooops your overrides have been overridden**

Ok, apart from this case also the core itself provides overrides https://github.com/joomla/joomla-cms/tree/4.0-dev/administrator/templates/atum/html which **CANNOT** be overridden by the user. So, to cut the story sort here, there are numerous **MAJOR** problems with the current templates:
- User is not able to actually override everything (as the project is advertising)
- There is no way any template to provide safe, self owned overrides and also allow the user to use/create their own ones
- Templates behave differently than the rest extensions by placing their static assets in subfolders like css/js/images inside them. The rest of the extensions are using the media folder. Now this might sound weird but there are couple of things with this: security and consistency.

## The solution
Provide child templates a-la wp.

### Summary of Changes

#### Architectural decisions
The table template_styles gets 2 more fields, named `parent` and `inherit`. Both fields default to 0 which translates to the current behaviour. The field parent is a boolean and acts as the enabler for the new functionality and also as a preventer if you try to make a child out of a child template. You can only inherit from a real template. the field parent should be exposed to the manifest as `<parent>1</parent>` for the templates allowing childs.
The field inherit should **not** be exposed in the manifest, should be treated as reserved. The value in that column is populated by com_templates whenever a child is created. The value is an integer and eqautes to the template style id that the child used as a parent.

There are changes in the View of the MVC, so that the template will search for the given template then if not found and it's a child in the parent then the system
There are changes in all the SiteApplication, AdministratorApplication that basically do the same thing as the procedure in the MVC
Same goes for the HTMLHelper that the includesFile is patched to cascade in the right folders
And finally the modules Helper to pick the correct override (child/parent)

The templates that implement the new mode **should have their static assets in the media folder either in `templates/site/template_name` or `templates/admin/template_name`**

These changes were all that was needed in the core system for this new mode to be enabled

### Testing Instructions
Download the intalling package at the end in the checks area (there is a download button) and install a copy of Joomla.

The installer should work as before.

Install the sampledata

Go to `/index.php?option=com_templates&view=templates&client_id=0` and click on the cassiopeia link. There is a new button `Create Child` click it. You should get a newly child template with the name that you gave in the modal. Click on close to go back and then click on the newly created template. The folder should have only an xml file and 2 images. Great!

Go to the site template styles annd select the newly child template. Make the template default annd then go to the menu assignement and reassing all menus to this template. Check the front end.
Wow a template without any files...

Create some overrides, component, module, layout in the Cassiopeia template
Edit each of the overridden file and add some simple text, so you can understand which file was used for the rendering of the page
Check the front end that although you were editing the parent template the child got them as well. Looking good...

Repeat the creation of the same overrides but this time on the newly created child template. Put some text to idendify the changes.

That's all


#### What is not there yet
- The com_templates is just monkey patched atm this can be demoed. It needs proper DRY code
- The creation of the overrides is not aware of the parent template and always using the com/mod/etc
- Patch the override notifier so it picks changes for both the parent template and the ext
- Some conditionals need some clean up for consistency

### Performance impact
Should be unnoticeable. The reason is that the core changes are using the template object (derives from the DB) which is cached, eg generated only once per lifecycle, and the additional payload are just few more conditionals. Static assets that have an option `relative => true` also get a couple more file lookups if the template is a child. All and all performance shouldn't be a negative factor here.


@wilsonge @mbabker @laoneo @ciar4n @C-Lodder can you some feedback here?

### Documentation Changes Required

Yes!
